aleph/logic/api_keys.py

```diff
@@ -0,0 +1,108 @@
+import datetime
+
+from flask import render_template
+from sqlalchemy import and_, or_, func
+
+from aleph.core import db
+from aleph.model import Role
+from aleph.model.common import make_token
+from aleph.logic.mail import email_role
+from aleph.logic.roles import update_role
+from aleph.logic.util import ui_url
+
+# Number of days after which API keys expire
+API_KEY_EXPIRATION_DAYS = 90
+
+# Number of days before an API key expires
+API_KEY_EXPIRES_SOON_DAYS = 7
+
+
+def generate_user_api_key(role):
+    event = "regenerated" if role.has_api_key else "generated"
+    params = {"role": role, "event": event}
+    plain = render_template("email/api_key_generated.txt", **params)
+    html = render_template("email/api_key_generated.html", **params)
+    subject = f"API key {event}"
+    email_role(role, subject, html=html, plain=plain)
+
+    now = datetime.datetime.utcnow().replace(microsecond=0)
+    role.api_key = make_token()
+    role.api_key_expires_at = now + datetime.timedelta(days=API_KEY_EXPIRATION_DAYS)
+    role.api_key_expiration_notification_sent = None
+
+    db.session.add(role)
+    db.session.commit()
+    update_role(role)
+
+    return role.api_key
+
+
+def send_api_key_expiration_notifications():
+    _send_api_key_expiration_notification(
+        days=7,
+        subject="Your API key will expire in 7 days",
+        plain_template="email/api_key_expires_soon.txt",
+        html_template="email/api_key_expires_soon.html",
+    )
+
+    _send_api_key_expiration_notification(
+        days=0,
+        subject="Your API key has expired",
+        plain_template="email/api_key_expired.txt",
+        html_template="email/api_key_expired.html",
+    )
+
+
+def _send_api_key_expiration_notification(
+    days,
+    subject,
+    plain_template,
+    html_template,
+):
+    now = datetime.date.today()
+    threshold = now + datetime.timedelta(days=days)
+
+    query = Role.all_users()
+    query = query.yield_per(1000)
+    query = query.where(
+        and_(
+            and_(
+                Role.api_key != None,  # noqa: E711
+                func.date(Role.api_key_expires_at) <= threshold,
+            ),
+            or_(
+                Role.api_key_expiration_notification_sent == None,  # noqa: E711
+                Role.api_key_expiration_notification_sent > days,
+            ),
+        )
+    )
+
+    for role in query:
+        params = {
+            "role": role,
+            "expires_at": role.api_key_expires_at,
+            "settings_url": ui_url("settings"),
+        }
+        plain = render_template(plain_template, **params)
+        html = render_template(html_template, **params)
+        email_role(role, subject, html=html, plain=plain)
+
+    query.update({Role.api_key_expiration_notification_sent: days})
+    db.session.commit()
+
+
+def reset_api_key_expiration():
```

Comment:
This is a good point! There are two aspects:

1. From a technical perspective, after rolling out this change, the API keys of all users will expire at the same time, and thus they will also all receive a notification at the same time. We’re already sending notification digests via email on a daily basis, and it usually takes approx. 1min per 1,500 users. Based on this and our total number of users, I figured it would probably be fine from a technical perspective. The only unknown here would be potential daily sending limits of our email provider.

   I had initially implemented a jitter of 0 to 20 days for the API key expiration, so that the expiration notifications where distributed across 20 days rather then sent out all on a single day. It added some complexity, and I initially didn’t feel that it’s worth it. But I will add it again just to be on the safe side.

2. From a user support perspective, I was planning to inform our internal teams that usually handle support requests from users so they are aware of this change and to display a dismissible message banner. I also tried to proactively avoid confusion by adding some context to the notification emails. Nonetheless, there will be some users who are confused by this. I’m honestly not sure if it will be helpful in this regard or cause more confusion if we send out the notification emails over a longer period or to send them out over multiple days. Do you have an opinion?

In any case, you are correct that this will be a one-time thing. Once legacy API keys have expired, users will have to manually regenerate keys if they want to keep using the API. This will automatically cause new expiration dates to vary depending on when users regenerate their keys. Also, it’s likely that only a small subset of users actually regenerates their keys.

Reply:
Thanks for the response and the insights. Don't introduce complexity if you don't feel it is required. I trust your judgement on this, and it's enough for me that you've evidently thought things though.